2-repositories/scala-repo-server/src/main/scala/com.snowplowanalytics.iglu.server/service/SchemaService.scala

```diff
@@ -45,96 +49,100 @@ import javax.ws.rs.Path
  * @param apiKeyActor a reference to a ``ApiKeyActor``
  */
-@Api(value = "/api/schemas", position = 0,
-  description = "Operations dealing with individual and multiple schemas")
+@Api(value = "/api/schemas", tags = Array("schema"),
+    authorizations = Array(new Authorization(value = "APIKeyHeader")), produces = "text/plain")
+@Path("/api/schemas")
 class SchemaService(schemaActor: ActorRef, apiKeyActor: ActorRef)
-(implicit executionContext: ExecutionContext) extends Directives with Service {
-
-  /**
-   * Creates a ``TokenAuthenticator`` to extract the apikey http header and
-   * validates it against the database.
-   */
-  val authenticator = TokenAuthenticator[(String, String)]("apikey") {
-    key => (apiKeyActor ? GetKey(key)).mapTo[Option[(String, String)]]
-  }
+                   (implicit executionContext: ExecutionContext, routingSettings: RoutingSettings)
+                   extends Directives with Service {
 
   /**
    * Directive to authenticate a user.
    */
-  def auth: Directive1[(String, String)] = authenticate(authenticator)
+  def auth(key: String): Directive1[(String, String)] = {
+    val credentialsRequest = (apiKeyActor ? GetKey(key)).mapTo[Option[(String, String)]] map {
+      // if the authentication succeeds
+      case Some(t) => Right(t)
+      // if the authentication fails
+      case None => Left(AuthenticationFailedRejection(CredentialsRejected, HttpChallenges.basic("")))
+    }
+    onSuccess(credentialsRequest).flatMap {
+      case Right(user) => provide(user)
+      case Left(rejection) => reject(rejection)
+    }
+  }
 
   /**
    * Directive to validate the schema provided (either by query param or form
    * data) is self-describing.
    */
   def validateSchema(format: String): Directive1[String] =
-    anyParam('schema) | entity(as[String]) flatMap { schema =>
+    formField('schema) | parameter('schema) | entity(as[String]) flatMap { schema =>
       onSuccess((schemaActor ? ValidateSchema(schema, format))
-          .mapTo[(StatusCode, String)]) flatMap { ext =>
-            ext match {
+          .mapTo[(StatusCode, String)]) tflatMap  {
               case (OK, j) => provide(j)
-              case res => complete(res)
+              case rej => complete(rej)
             }
       }
-    }
 
   /**
    * Schema service's route
    */
-  lazy val routes =
+  lazy val routes: Route =
     rejectEmptyResponse {
-      respondWithMediaType(`application/json`) {
-        auth { authPair =>
-          post {
-            addRoute(authPair._1, authPair._2)
-          } ~
-          put {
-            updateRoute(authPair._1, authPair._2)
-          } ~
-          delete {
-            deleteRoute(authPair._1, authPair._2)
-          } ~
-          get {
-            path("public") {
-              publicSchemasRoute(authPair._1, authPair._2)
+      (get | post | put | delete) {
+        headerValueByName("apikey") { apikey =>
+          auth(apikey) { authPair =>
```

Comment:
This should be `auth(apikey) { case (owner, permission) =>` to avoid all `_._1` and `_._2`

Reply:
Done